File: 2020/Java/src/aoc/Day6.java
package aoc;

import java.io.IOException;
import java.nio.file.Files;
import java.util.HashSet;
import java.util.Set;

public abstract class Day6 {

	public static void main(String[] args) throws IOException {
		int globalGroupCount = 0;
		int globalCount = 0;
		Set<Byte> groupFoundAny = new HashSet<>();
		Set<Byte> groupFoundEvery = new HashSet<>();
		Set<Byte> found = new HashSet<>();
		boolean newline = false;
		boolean first = true;
		for (byte b : Files.readAllBytes(InputReader.getInputFile(6).toPath())) {
			if (b == '\n') {
				if (newline) {
					globalCount += groupFoundAny.size();
					globalGroupCount += groupFoundEvery.size();
					groupFoundAny.clear();
					groupFoundEvery.clear();
					first = true;
				} else {
					groupFoundAny.addAll(found);
					if (first) {
						groupFoundEvery.addAll(found);
						first = false;
					} else {
						groupFoundEvery.retainAll(found);
					}
					found.clear();
					newline = true;
				}
			} else if (b == '\r') {
			} else {
				newline = false;
				found.add(b);
			}
		}

		globalCount += groupFoundAny.size();
		globalGroupCount += groupFoundEvery.size();

		System.out.printf("The sum of all group part 1 answers is %d.%n", globalCount);
		System.out.printf("The sum of all group part 2 answers is %d.%n", globalGroupCount);
	}
}

File: 2020/Python/Day6.py
import os
import sys


def main():
    inf = os.path.dirname(os.getcwd())
    inf = os.path.join(inf, "input")
    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf += ".txt"
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print(inf + " does not exist!")
        inf = os.path.join(os.path.dirname(inf), "input")
        inf = os.path.join(inf, "Day6.txt")
        print(f"trying {inf} instead.")

    if not os.path.exists(inf) or os.path.isdir(inf):
        print("None of the expected inputs exist!", file=sys.stderr)
        return
    else:
        print(f"Using input file {inf}.")

    inputFile = open(inf, 'r')
    globalGroupFound = 0
    globalFound = 0
    groupFoundEvery = []
    groupFoundAny = []
    found = []
    newline = False
    first = True
    for b in inputFile.read():
        if b == '\n':
            if newline:
                globalFound += len(groupFoundAny)
                globalGroupFound += len(groupFoundEvery)
                first = True
            else:
                if first:
                    groupFoundAny = found.copy()
                    groupFoundEvery = found.copy()
                    first = False
                else:
                    for a in found:
                        if not a in groupFoundAny:
                            groupFoundAny.append(a)
                    groupFoundEvery = list(set(groupFoundEvery) & set(found))

                found.clear()
                newline = True
        else:
            newline = False
            if not b in found:
                found.append(b)

    globalFound += len(groupFoundAny)
    globalGroupFound += len(groupFoundEvery)

    print(f"The sum of all group part 1 answers is {globalFound}.")
    print(f"The sum of all group part 2 answers is {globalGroupFound}.")


if __name__ == '__main__':
    main()
